Development Documentation Overhaul

mdhaber opened the pull request:
Closes #8353
Closes #8966
Closes #9951
Closes #10355
Partially addresses #8575, but information about using Cython as a bridge to other compiled code should be added.
`development_workflow.rst` and all files in the `/gitwash` adapted from [NumPy documentation](https://github.com/numpy/numpy/tree/master/doc/source/dev).

Most existing SciPy documentation has been preserved, but may have been moved (e.g. parts from `HACKING.rst.txt` have been redistributed). Please see self-review for more information.

Questions:
- [x] Why does HACKING.rst.tst exist separately in the `scipy` root directory? Why not just put the contents in `\doc\source\hacking.rst`? Just curious.
- [x] The extra slow tests are run by adding ``SCIPY_XSLOW=1`` as an OS environment variable?
- [x] How should I credit NumPy for the Git and Development Workflow documentation?

TODO:
- [x] Fix CI failure.
- [x] link to CONTRIBUTING.rst somewhere. (I don't know that I've run across this before, so if it had a link before perhaps it was buried?)
- [x] Read through everything. I know there are a lot of issues due to things moving around.

mdhaber commented on HACKING.rst.txt:
```diff
@@ -194,289 +219,6 @@ project and SciPy the community, and it can always use a new pair of hands.
-
-
-*Are there any video examples for installing from source, setting up a
-development environment, etc...?*
```
Removed question. Link is in new [macOS Quickstart guide](https://github.com/scipy/scipy/pull/10286/files#diff-23d530dd47e20cba4296e7e42f36ac9c).

mdhaber commented on HACKING.rst.txt:
```diff
@@ -194,289 +219,6 @@ project and SciPy the community, and it can always use a new pair of hands.
-
-
-*Are there any video examples of the basic development workflow?*
-
```
Link is in CTOC. Should probably be moved to [Development Workflow](https://github.com/scipy/scipy/pull/10286/files#diff-b0195965dcec36eb5061a427eb5c5da4) document.

mdhaber commented on HACKING.rst.txt:
```diff
@@ -194,289 +219,6 @@ project and SciPy the community, and it can always use a new pair of hands.
-
-
-*Can I use a programming language other than Python to speed up my code?*
-
```
Moved to [`doc\source\dev\contributor\compiled_code.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-0277b675f933e117e69eaf469d0dbb37) linked from CTOC.

mdhaber commented on HACKING.rst.txt:
```diff
@@ -194,289 +219,6 @@ project and SciPy the community, and it can always use a new pair of hands.
-please discuss those reasons first.
-
-
-*How do I debug code written in C/C++/Fortran inside Scipy?*
```
Moved to [`doc\source\dev\contributor\compiled_code.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-0277b675f933e117e69eaf469d0dbb37) linked from CTOC.

mdhaber commented on HACKING.rst.txt:
```diff
@@ -194,289 +219,6 @@ project and SciPy the community, and it can always use a new pair of hands.
-``python`` binary with arguments ``runtests.py -g --python mytest.py``.
-
-
-*How do I enable additional tests in Scipy?*
```
Information added to new [`doc\source\dev\contributor\runtests.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-2dd64dc534d9aec2139341f236d14faf) linked from CTOC.

mdhaber commented on doc/API.rst.txt:
```diff
@@ -138,3 +146,47 @@ change is made.
+
+  
+SciPy structure
+---------------
```
Moved from [HACKING.rst.txt](https://github.com/scipy/scipy/pull/10286/files#diff-cb78383d096ab473e4e752dc818fcfda)

mdhaber commented on doc/source/building/faq.rst:
```diff
@@ -0,0 +1,51 @@
+
+=================
+Build/Install FAQ
+=================
```
Moved from [HACKING.rst.txt](https://github.com/scipy/scipy/pull/10286/files#diff-cb78383d096ab473e4e752dc818fcfda)

mdhaber commented on doc/source/building/index.rst:
```diff
@@ -16,4 +16,5 @@ Build instructions for different operating systems:
    macosx
-
+   
+For additional information, see the :ref:`build-install-faq`.
```
Links to [Build/Install FAQ](https://github.com/scipy/scipy/pull/10286/files#diff-ff7a395fc10e317f57cd03639d848aeb)

mdhaber commented on doc/source/dev/contributor/benchmarking.rst:
```diff
@@ -0,0 +1,208 @@
+.. _benchmarking-with-asv:
+
+Benchmarking SciPy with Airspeed Velocity
+=========================================
```
New. Contains most of what `benchmarks\README.rst` (not linked from documentation) has and much more.

mdhaber commented on doc/source/dev/contributor/compiled_code.rst:
```diff
@@ -0,0 +1,51 @@
+
+=============
+Beyond Python
+=============
```
Moved from [HACKING.rst.txt](https://github.com/scipy/scipy/pull/10286/files#diff-cb78383d096ab473e4e752dc818fcfda). Linked from CTOC.

mdhaber commented on doc/source/dev/contributor/contributor_toc.rst:
```diff
@@ -0,0 +1,76 @@
+
+=============================
+Contributor Table of Contents
+=============================
```
New. Links to everything useful to new contributor that I could find. Linked from [`/source/index.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-4bc0dbe7e164d778c3dc17f70c9f15ac)

mdhaber commented on doc/source/dev/contributor/cython.rst:
```diff
@@ -0,0 +1,202 @@
+.. _adding-cython:
+
+Adding Cython to SciPy
+======================
```
New. Linked from CTOC. Could use section about interfacing with other compiled code.

mdhaber commented on doc/source/dev/contributor/development_workflow.rst:
```diff
@@ -0,0 +1,228 @@
+
+====================
+Development workflow
+====================
```
Adapted from NumPy. How should I give them credit?

mdhaber commented on doc/source/dev/contributor/development_workflow.rst:
```diff
@@ -0,0 +1,228 @@
+
+.. _pr-checklist:
+
+Checklist before submitting a PR
```
Moved from `HACKING.rst.txt`.

mdhaber commented on doc/source/dev/contributor/pep8.rst:
```diff
@@ -0,0 +1,52 @@
+
+==============
+PEP8 and SciPy
+==============
```
New.

mdhaber commented on doc/source/dev/contributor/quickstart_mac.rst:
```diff
@@ -0,0 +1,109 @@
+
+================================================
+Development Environment Quickstart Quide (macOS)
+================================================
```
New. Linked only from CTOC right now, but basically is a more detailed version of the macOS instructions (`doc/source/building/macosx.rst`) linked from `doc/source/building/index.rst`. For now, I'm leaving these both in the documentation. CTOC leads new contributors to this "Quickstart Guide" if they have macOS and to the other build docs otherwise. I will consider replacing `doc/source/building/macosx.rst` with this someday, but probably only after I can also replace `doc/source/building/linux.rst`.

mdhaber commented on doc/source/dev/contributor/recommended_development_setup.rst:
```diff
@@ -0,0 +1,54 @@
+
+=============================
+Recommended development setup
+=============================
```
Wasn't sure where to put this, so it's just a separate document linked from CTOC. All of this information is in a more complete setting in the [macOS Quickstart](https://github.com/scipy/scipy/pull/10286/files#diff-23d530dd47e20cba4296e7e42f36ac9c); this might still be needed for Linux/Windows users. It probably should be integrated into those instructions, but I am not modifying those in this PR.

mdhaber commented on doc/source/dev/contributor/recommended_development_setup.rst:
```diff
@@ -0,0 +1,54 @@
+build or installing to a virtualenv. See the FAQ below for details.
+
+Some of the tests in Scipy are very slow and need to be separately
+enabled. See the FAQ below for details.
```
> See the FAQ below for details.

I'll be reading to remove outdated stuff like this before merge.

mdhaber commented on doc/source/dev/contributor/runtests.rst:
```diff
@@ -0,0 +1,112 @@
+
+===========================
+Running SciPy Tests Locally
+===========================
```
New.

mdhaber commented on doc/source/dev/gitwash/configure_git.rst:
```diff
@@ -0,0 +1,123 @@
+
+=================
+Git configuration
+=================
```
This and the rest of `doc/source/dev/gitwash` adapted from NumPy documentation.

mdhaber commented on doc/source/dev/gitwash/gitwash.rst:
```diff
@@ -0,0 +1,34 @@
+
+=====================
+ Git for development
+=====================
```
NumPy removed the Development Workflow documentation from this list (https://www.numpy.org/devdocs/dev/gitwash/index.html) so I followed suit. It feels sort of incomplete without it, though.

mdhaber commented on doc/source/dev/gitwash/useful_git.rst:
```diff
@@ -0,0 +1,309 @@
+
+========
+Git tips
+========
```
Moved from Development Workflow.

mdhaber commented on doc/source/dev/index.rst:
```diff
@@ -1,6 +1,8 @@
+
+==========================
+SciPy Core Developer Guide
+==========================
```
As a new contributor, wouldn't you think you would want to _start_ with the "SciPy Developer Guide"? Most of this is really appropriate for new contributors, though. Some of it could be useful, so it's linked from the CTOC, but hopefully now this looks less like it's supposed to be starting point for new contributors.

mdhaber commented:
CI failure presumably for:
```
API.rst.txt:175: WARNING: Unknown target name: "how to document".
/home/circleci/repo/doc/source/building/faq.rst:10: WARNING: Unknown target name: "virtualenv".
/home/circleci/repo/doc/source/building/faq.rst:10: WARNING: Unknown target name: "virtualenvwrapper".
/home/circleci/repo/doc/source/building/faq.rst:39: WARNING: Unknown target name: "spyder".
/home/circleci/repo/doc/source/dev/contributor/quickstart_mac.rst:59: WARNING: Title underline too short.

Installing SciPy
--------------
/home/circleci/repo/doc/source/dev/contributor/quickstart_mac.rst:59: WARNING: Title underline too short.

Installing SciPy
--------------
/home/circleci/repo/doc/source/dev/contributor/quickstart_mac.rst:5: WARNING: Duplicate explicit target name: "here".
/home/circleci/repo/doc/source/dev/gitwash/useful_git.rst:227: WARNING: Unknown target name: "scipy/master".
```
Will fix.

rgommers commented:
> Why does HACKING.rst.tst exist separately in the scipy root directory? Why not just put the contents in \doc\source\hacking.rst? Just curious.

It predates more extensive html docs, and that file name is (used to be) commonly used by projects for developer docs.

> The extra slow tests are run by adding `SCIPY_XSLOW=1` as an OS environment variable?

yes

> How should I credit NumPy for the Git and Development Workflow documentation?

don't, in the docs themselves. just put it in the description of this PR. that applies to other credits too (e.g. I saw a "thanks to Yarik ..." note somewhere).

> link to CONTRIBUTING.rst somewhere. (I don't know that I've run across this before, so if it had a link before perhaps it was buried?)

I don't think you should link back to that. It's a very concise intro, and is a file that GitHub recognizes so it needs to be there. It's an intro point to link to the real docs, no need to reverse that flow of information/attention.

> Contributing to SciPy
> Contributor Table of Contents

This naming is a bit odd. How about something like "Ways to contribute to SciPy" and "Contributor Guide"?

mdhaber commented:
Re: "Thanks to Yarik", I didn't add that. It was in the NumPy version, but I'll remove it.
Yes, I'll change the names.

mdhaber commented:
Changed names and fixed some errors, but somehow my changes are making Sphinx complain about 100 files that aren't included in any toctree. I compared to another doc build, and these warnings aren't present there, so I assume these are why `Makefile:115: recipe for target 'html-scipyorg' failed`. Some of them are for my files, but most of them aren't. How could the changes I made to the development documentation break all of this content generated from docstrings?

Does everything have to be in a toctree?

```
WARNING: duplicate object description of scipy.io.wavfile, other instance in io, use :noindex: for one of them
WARNING: duplicate object description of scipy.io.arff, other instance in io, use :noindex: for one of them
WARNING: duplicate object description of scipy.io, other instance in io, use :noindex: for one of them
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/circleci/repo/doc/source/building/faq.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/building/index.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/contributor/benchmarking.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/contributor/compiled_code.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/contributor/cython.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/contributor/development_workflow.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/contributor/pep8.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/contributor/quickstart_mac.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/contributor/recommended_development_setup.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/contributor/runtests.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/dev/gitwash/gitwash.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.function.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.signature.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.LowLevelCallable.user_data.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.integrate.BDF.step_size.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.integrate.LSODA.step_size.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.integrate.OdeSolver.step_size.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.integrate.RK23.step_size.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.integrate.RK45.step_size.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.integrate.Radau.step_size.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.integrate.complex_ode.y.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.integrate.ode.y.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.interpolate.BSpline.tck.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.interpolate.Rbf.A.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.interpolate.interp1d.fill_value.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.io.netcdf_variable.isrec.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.io.netcdf_variable.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.optimize.LbfgsInvHessProduct.H.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.optimize.LbfgsInvHessProduct.T.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.StateSpace.A.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.StateSpace.B.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.StateSpace.C.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.StateSpace.D.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.StateSpace.dt.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.StateSpace.poles.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.StateSpace.zeros.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.TransferFunction.den.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.TransferFunction.dt.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.TransferFunction.num.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.TransferFunction.poles.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.TransferFunction.zeros.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.ZerosPolesGain.dt.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.ZerosPolesGain.gain.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.ZerosPolesGain.poles.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.ZerosPolesGain.zeros.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.dlti.dt.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.dlti.poles.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.dlti.zeros.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.lti.dt.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.lti.poles.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.signal.lti.zeros.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.bsr_matrix.blocksize.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.bsr_matrix.dtype.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.bsr_matrix.has_canonical_format.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.bsr_matrix.has_sorted_indices.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.bsr_matrix.nnz.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.bsr_matrix.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.coo_matrix.dtype.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.coo_matrix.nnz.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.coo_matrix.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csc_matrix.dtype.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csc_matrix.has_canonical_format.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csc_matrix.has_sorted_indices.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csc_matrix.nnz.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csc_matrix.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csr_matrix.dtype.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csr_matrix.has_canonical_format.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csr_matrix.has_sorted_indices.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csr_matrix.nnz.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.csr_matrix.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.dia_matrix.dtype.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.dia_matrix.nnz.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.dia_matrix.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.dok_matrix.nnz.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.dok_matrix.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.lil_matrix.nnz.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.lil_matrix.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.linalg.LinearOperator.H.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.linalg.LinearOperator.T.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.spmatrix.nnz.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.sparse.spmatrix.shape.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.ConvexHull.points.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.ConvexHull.vertices.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.Delaunay.convex_hull.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.Delaunay.points.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.Delaunay.transform.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.Delaunay.vertex_neighbor_vertices.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.Delaunay.vertex_to_simplex.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.HalfspaceIntersection.dual_vertices.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.HalfspaceIntersection.halfspaces.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.Voronoi.points.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.spatial.Voronoi.ridge_dict.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.stats.gaussian_kde.neff.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.stats.gaussian_kde.weights.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.stats.rv_continuous.random_state.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.stats.rv_discrete.random_state.rst: WARNING: document isn't included in any toctree
/home/circleci/repo/doc/source/generated/scipy.stats.rv_histogram.random_state.rst: WARNING: document isn't included in any toctree
```

mdhaber commented:
I'm trying to improve the checklist for submitting a PR:
https://13487-1460385-gh.circle-artifacts.com/0/html-scipyorg/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

* What is the easiest way to find the version number of the next release, if not running `scipy.__version__` from the development version? I don't see it in `setup.py`. 
* Is the wiki the place to add to the release notes? Is that a reliable way to get the next version number?

mdhaber commented:
@larsoner Hi Eric, you mentioned you'd be willing to review this. Before that, I was wondering if you had any idea what is causing the Sphinx build to fail. My comment above:

> Changed names and fixed some errors...

describes the issue in more detail. I see there are also some new errors:

```
/home/circleci/repo/doc/source/dev/contributor/rendering_documentation.rst:44: WARNING: malformed hyperlink target.
source/dev/core-dev/github.rst:10: WARNING: duplicate label labels-and-milestones, other instance in /home/circleci/repo/doc/source/dev/core-dev/github.rst
source/dev/core-dev/github.rst:61: WARNING: duplicate label backporting, other instance in /home/circleci/repo/doc/source/dev/core-dev/github.rst
source/dev/core-dev/deprecations.rst:4: WARNING: duplicate label deprecations, other instance in /home/circleci/repo/doc/source/dev/core-dev/deprecations.rst
source/dev/core-dev/distributing.rst:4: WARNING: duplicate label distributing-a-release, other instance in /home/circleci/repo/doc/source/dev/core-dev/distributing.rst
source/dev/core-dev/distributing.rst:88: WARNING: duplicate label supported-py-numpy-versions, other instance in /home/circleci/repo/doc/source/dev/core-dev/distributing.rst
/home/circleci/repo/doc/source/dev/core-dev/releasing.rst:4: WARNING: duplicate label making-a-release, other instance in /home/circleci/repo/doc/source/dev/core-dev/index.rst
/home/circleci/repo/doc/source/dev/core-dev/versioning.rst:4: WARNING: duplicate label version-numbering, other instance in /home/circleci/repo/doc/source/dev/core-dev/index.rst
```
I understand the first one about the "malformed hyperlink target". But the rest about duplicate labels are puzzling. They just showed up when I moved the Core Developer documentation files into their own folder. I didn't make changes to all those files, and there are not duplicate labels within those documents.

There is also:
```
../HACKING.rst.txt:26: WARNING: undefined label: deciding-on-new-features (if the link has no caption the label must precede a section header)
```
Which I understand. I actually had defined the label, then removed it when I saw Sphinx complaining about duplicate labels locally.

If you can't think of anything, I might need to just re-create this branch and PR. I guess I'd start with a fresh master and re-make changes, proceeding carefully to determine when the errors start to pop up.

larsoner commented:
Should be able to look tomorrow. In the meantime can you rebase on latest `master` to make sure it hasn't been fixed there?

mdhaber commented:
@larsoner Genius! (or I'm a noob.) 
I'm sure I rebased on master before branching, so I didn't suspect there was something wrong with it. 

Now all that's left is the duplicate labels:
```
source/dev/core-dev/github.rst:10: WARNING: duplicate label labels-and-milestones, other instance in /home/circleci/repo/doc/source/dev/core-dev/github.rst
source/dev/core-dev/github.rst:61: WARNING: duplicate label backporting, other instance in /home/circleci/repo/doc/source/dev/core-dev/github.rst
source/dev/core-dev/deprecations.rst:4: WARNING: duplicate label deprecations, other instance in /home/circleci/repo/doc/source/dev/core-dev/deprecations.rst
source/dev/core-dev/distributing.rst:4: WARNING: duplicate label distributing-a-release, other instance in /home/circleci/repo/doc/source/dev/core-dev/distributing.rst
source/dev/core-dev/distributing.rst:88: WARNING: duplicate label supported-py-numpy-versions, other instance in /home/circleci/repo/doc/source/dev/core-dev/distributing.rst
/home/circleci/repo/doc/source/dev/core-dev/newfeatures.rst:5: WARNING: duplicate label deciding-on-new-features, other instance in /home/circleci/repo/doc/source/dev/core-dev/index.rst
/home/circleci/repo/doc/source/dev/core-dev/releasing.rst:4: WARNING: duplicate label making-a-release, other instance in /home/circleci/repo/doc/source/dev/core-dev/index.rst
/home/circleci/repo/doc/source/dev/core-dev/versioning.rst:4: WARNING: duplicate label version-numbering, other instance in /home/circleci/repo/doc/source/dev/core-dev/index.rst
```
For your convenience, links to the "offending" files:
[`github.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-4f87b120208753f39fbe0968531906fb)
[`deprecations.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-9ee51ad5cab99e2276608030e19697a8)
[`distributing.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-73c5270ff9d8bee2470513e42e808d5d)
[`newfeatures.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-1206a1ddb2698b10a112b541996dcc3b)
[`releasing.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-aba6b5f1e6f777df1ef716f3291a7485)
[`versioning.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-a9086c8fc08d70f5449f456bac3ad1a7)
[`index.rst`](https://github.com/scipy/scipy/pull/10286/files#diff-4d294dac041230c70f8e61198b7b683c)

fletcheaston commented on doc/source/dev/contributor/quickstart_mac.rst:
```diff
@@ -0,0 +1,109 @@
+
+#. (Optional) Check your present working directory by entering ``pwd`` at the terminal. You should be in the root ``/scipy`` directory, not in a directory ending ``/scipy/scipy``.
+
+#. Do an in-place build: enter ``python3 setup.py build_ext --inplace``. |br| This will compile the C, C++, and Fortran code that comes with SciPy. We installed ``python3`` with Anaconda. ``setup.py`` is a script in the root directory of SciPy, which is why you have to be in the SciPy root directory to call it. ``build_ext`` is a command defined in ``setup.py``, and ``--inplace`` is an option we' ...
```
Explaining what the command does and what the flags do is a good move, especially for those who haven't built SciPy or similar libraries before.

fletcheaston commented on doc/source/dev/contributor/quickstart_mac.rst:
```diff
@@ -0,0 +1,109 @@
+   
+   - ``unset PYTHONPATH``,
+   
+   respectively.
```
At first, it seemed a bit unclear that the ~/activate.d/env_vars.sh file needed the "export PYTHONPATH=/scipy" line, and that the ~/deactivate.d/env_vars.sh file needed the "unset PYTHONPATH" line. Originally, I put both lines into both files, and only remedied this after viewing the videos. Making it more clear ahead of time (as opposed to adding "respectively" after the fact) which line goes in which file would be helpful.

mdhaber replied:
Noted.

fletcheaston replied:
I don't believe a fix for this is necessary, if we use the conda develop instead.

mdhaber replied:
OK. The thing is that I can't easily fix the video. I was thinking maybe offering both options?

fletcheaston replied:
Makes sense. I'll rewrite it and offer both options.

mdhaber commented on doc/source/dev/contributor/quickstart_mac.rst:
```diff
@@ -0,0 +1,109 @@
+
+#. Enter ``conda install cython numpy matplotlib pytest spyder``. |br| Note that we're only installing SciPy's dependencies (and Spyder so we can use the IDE), but not SciPy itself.
+
+#. Our goal now is to add our root SciPy directory to the ``PYTHONPATH`` environment variable whenever this virtual environment is activated. This will ensure that Python can find the SciPy code we are trying to `import`. This requires adding a few files and folders deep inside your Anaconda installation directory. I suggest watching `the video <https://youtu.be/Faz29u5xIZc?t=35>`_ for this part. ...
```
Good find! I created the video about a year ago, maybe more. It looks like I made a typo when I (recently) typed the document based on the video. 
I knew about changing PYTHONPATH from within Spyder, and you can toggle that on and off, but it's not as convenient as the solution I chose to present (once you get it set up right). Anyway, I didn't find this `conda develop` solution then. I'll look into it.

mdhaber replied:
Can you write revised steps and indicate which ones would be replaced?

fletcheaston replied:
Definitely!

mdhaber commented on doc/source/dev/contributor/quickstart_mac.rst:
```diff
@@ -0,0 +1,109 @@
+
+#. Install `Homebrew`_. Enter into the terminal |br| ``/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"`` |br| or follow the installation instructions listed on the Homebrew website. Homebrew is a package manager for macOS that will help you download `gcc`, the software we will use to compile C, C++, and Fortran code included in SciPy.
+
+#. Use Homebrew to install ``gcc`` by entering the command ``brew install gcc``.
```
Great to hear!

mdhaber commented on doc/source/dev/contributor/quickstart_mac.rst:
```diff
@@ -0,0 +1,109 @@
+* performing an in-place build of SciPy; and 
+* creating a virtual environment that adds this development version of SciPy to the Python path on macOS.
+
+Its companion videos `Anaconda SciPy Dev: Part I (macOS)`_ and `Anaconda SciPy Dev: Part I (macOS)`_ show many of the steps being performed. This guide may diverge slightly from the videos over time with the goal of keeping this guide the simplest, up-to-date procedure.
```
Great.

larsoner commented on doc/source/dev/contributor/contributor_toc.rst:
```diff
@@ -0,0 +1,94 @@
+- :ref:`adding-cython` - Extending and compiling Python code with `Cython`_ can significantly improve its performance. This document helps you get started.
+- :ref:`other-languages` discusses the use of C, C++, and Fortran code in SciPy.
+
+Here is a toctree, purely to appease Sphinx.
```
You can use `:orphan:` at the top of the files that do not belong to a toctree / that you link to manually.

Let me know when you push and CircleCI is happy and I'll look

mdhaber replied:
Done

fletcheaston commented on doc/source/dev/contributor/rendering_documentation.rst:
```diff
@@ -0,0 +1,72 @@
+
+To render the documentation on your own machine (macOS or Linux):
+
+#. Install `Sphinx`_ or ensure that your installation is up to date.
```
Ensuring that the user is in the correct development environment and providing the "conda update sphinx" command would make installing Sphinx very simple and easy. This step is also presented in the video walkthrough, as opposed to installing Sphinx through the linked website.

fletcheaston commented on doc/source/dev/contributor/rendering_documentation.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. note::
+
+   Changes to certain documents do not take effect when Sphinx documentation
```
Do we know what kinds of changes exhibit this behavior, or is it seemingly arbitrary?

mdhaber replied:
I know it happens when updating documentation of solver methods, but I'm not sure if that's the only time, so I didn't speculate. If someone else has an answer it would be good to know.

mdhaber commented:
@larsoner CI is happy and the toctrees are gone. I still need to make the modifications Fletcher suggested, but they are relatively small.

larsoner commented on doc/source/dev/contributor/rendering_documentation.rst:
```diff
@@ -0,0 +1,74 @@
+
+#. At the bottom of the PR, select “Show all Checks”.
+#. Next to “ci/circleci: build_docs”, select “Details”.
+#. Find the CircleCI job number. This is the 4-5 digit number in the URL of
```
This does not seem necessary. When I log in with a GitHub user with no SciPy privileges I see:

![Screenshot from 2019-06-24 15-01-44](https://user-images.githubusercontent.com/2365790/60044893-6a56ab80-9691-11e9-8fae-69db1ca6e76c.png)

Clicking the "Details" link takes me to:

![Screenshot from 2019-06-24 15-02-16](https://user-images.githubusercontent.com/2365790/60044987-a427b200-9691-11e9-9c33-1cac8bfd6736.png)

If it does not, you just need to sign in to CircleCI, authenticated with GitHub, and try again. No need for this manual stuff.

Then the last step is the standard "Artifacts" tab, where you can get to `index.html`.

mdhaber replied:
Done.

larsoner commented on doc/source/dev/contributor/rendering_documentation.rst:
```diff
@@ -0,0 +1,74 @@
+
+#. Log into CircleCI with your GitHub account.
+#. Visit the “ci/circleci: build_docs” job page as explained above.
+#. Select the “Artifacts” tab and find ``index.html``.
```
All of this can be removed, given the above.

FYI any GitHub user can also access https://circleci.com/gh/scipy to see all builds, but they probably don't care in most cases.

mdhaber replied:
Maybe that's new? At some point I think that URL trick was needed. Thanks.

larsoner commented on doc/source/dev/gitwash/development_setup.rst:
```diff
@@ -0,0 +1,143 @@
+You may also want to have easy access to all pull requests sent to the
+SciPy repository::
+
+   git config --add remote.upstream.fetch '+refs/pull/*/head:refs/remotes/upstream/pr/*'
```
This is a ton of PRs, I would not recommend this. I've tried it before and it's actually a bit of a nightmare for something like SciPy.

mdhaber replied:
This whole Git tutorial was brought over from NumPy documentation. Mostly I just converted "NumPy" to "SciPy". I think it's an improvement to have it, but it needs a lot of attention. I'll get rid of this part, though.

rgommers replied:
This used to be the best way to do it, but there's https://hub.github.com/ now which is way nicer.

mdhaber replied:
I didn't think just mentioning hub fit with the rest of the tutorial, so I removed this part entirely. I assumed this removes the line from `.git/config`:
```
fetch = +refs/pull/*/head:refs/remotes/upstream/pr/*
```

larsoner commented on doc/source/dev/gitwash/dot2_dot3.rst:
```diff
@@ -0,0 +1,26 @@
+
+========================================
+ Two and three dots in difference specs
+========================================
```
Is the `=` and title misalignment intentional here?

mdhaber replied:
I don't think that document really deserves to exist on its own, but for now I brought it in (from NumPy) anyway. Didn't notice this before, but I'll fix it until this information finds its way into somewhere appropriate.

tylerjereddy commented:
> What is the easiest way to find the version number of the next release, if not running scipy.__version__ from the development version? The checklist says to get it from setup.py, but I don't see it there. Update: I'm recommending getting it from the version number of the newest release notes on the wiki.

It is in `setup.py` (see below) for development branch, but you'd be right that a released version of SciPy probably wouldn't have the **next** version string around IIRC. I suppose you could also check the milestones available on GitHub, but programmatically probing the next SciPy stable release from an older stable release would indeed seem confusing to me.

```python
 56 MAJOR = 1
 57 MINOR = 4
 58 MICRO = 0
 59 ISRELEASED = False
 60 VERSION = '%d.%d.%d' % (MAJOR, MINOR, MICRO)
```

mdhaber commented:
Thanks @tylerjreddy didn't see that before. I'm adding it to the checklist.

mdhaber commented on doc/source/dev/contributor/recommended_development_setup.rst:
```diff
@@ -0,0 +1,54 @@
+explained below.
+
+First fork a copy of the main SciPy repository in Github onto your own
+account and then create your local repository via::
```
For now I did that to this document since it already has the `$` at the beginning of each line. Let's create a separate issue if it's important to add this to all of the documentation.

mdhaber commented on doc/source/dev/contributor/development_workflow.rst:
```diff
@@ -0,0 +1,309 @@
+   documentation of |differential_evolution|_, for example. You can get the
+   next version number from the most recent release notes on `the wiki`_.
+-  In case of larger additions, is there a tutorial or more extensive
+   module-level description? Tutorial files are in ``doc/source/tutorial``.
```
I thought about that. When do we require benchmarks? Should this be added to `HACKING.rst.txt`, too?

mdhaber replied:
Added this a few bullets up.

mdhaber commented on doc/source/building/faq.rst:
```diff
@@ -0,0 +1,63 @@
+
+Now editing a Python source file in SciPy allows you to immediately
+test and use your changes (in ``.py`` files), by simply restarting the
+interpreter.
```
I spent about a minute searching for this but I'm not finding clear answers.
Is that the complete command? 
What does it do, and what are the advantages?

larsoner replied:
> What does it do, and what are the advantages?

It installs `scipy` in your current environment in editable mode. This is the complete command that I use, except I append a `--user` (but this is not the recommended way to use python envs).

We can add this later if we want, since others might know if it's a bad idea.

mdhaber commented on doc/source/dev/contributor/benchmarking.rst:
```diff
@@ -0,0 +1,210 @@
+.. _Benchmarking SciPy: https://youtu.be/edLQ8KRpupQ
+.. _airspeed velocity (asv) documentation: https://asv.readthedocs.io/en/stable/
+.. _airspeed velocity of an unladen scipy: https://pv.github.io/scipy-bench/
+.. _Writing benchmarks: https://asv.readthedocs.io/en/stable/writing_benchmarks.html
```
As if working with Sphinx links wasn't fun enough already.

mdhaber replied:
Added to `conf.py`
```
intersphinx_mapping = {
    'python': ('https://docs.python.org/dev', None),
    'numpy': ('https://docs.scipy.org/doc/numpy', None),
    'matplotlib': ('https://matplotlib.org', None),
    'asv': ('https://asv.readthedocs.io/', None),
}
```
Added to `benchmarking.rst`:
```
:ref:`Writing benchmarks <asv:writing-benchmarks>`
```
Didn't work. By repeatedly editing whitespace elsewhere in the file it will probably fix itself eventually, but hasn't yet.

mdhaber replied:
`https://asv.readthedocs.io/en/stable/` FTW.

mdhaber replied:
Did the same for A Guide to NumPy/SciPy Documentation.
Tried to do the same for NumPy Distutils - Users Guide and NumPy/SciPy Testing Guidelines in the Contributor Guide TOC, but those documents don't have reference labels.

mdhaber commented on HACKING.rst.txt:
```diff
@@ -75,25 +80,25 @@ tests and correct code style.
     tabs by four spaces.  Using `pyflakes`_ to check your code is also a good
     idea.
```
OK. I'm going through this list from the bottom up and commented about this when you made a similar comment about the checklist.

mdhaber commented on doc/source/dev/contributor/contributor_toc.rst:
```diff
@@ -0,0 +1,82 @@
+
+.. _NumPy Distutils - Users Guide: https://docs.scipy.org/doc/numpy/reference/distutils_guide.html
+
+.. _NumPy/SciPy Testing Guidelines: https://docs.scipy.org/doc/numpy/reference/testing.html
```
Done.

mdhaber commented on doc/source/dev/contributor/contributor_toc.rst:
```diff
@@ -0,0 +1,82 @@
+
+.. _NumPy/SciPy Testing Guidelines: https://docs.scipy.org/doc/numpy/reference/testing.html
+
+.. _numpydoc docstring guide: https://numpydoc.readthedocs.io/en/latest/format.html
```
How about [:ref:`numpy:howto-document`](https://docs.scipy.org/doc/numpy/docs/howto_document.html#howto-document)?

mdhaber commented on doc/source/dev/contributor/contributor_toc.rst:
```diff
@@ -0,0 +1,82 @@
+
+.. _Scipy Development Workflow: https://youtu.be/HgU01gJbzMY
+
+.. _NumPy Distutils - Users Guide: https://docs.scipy.org/doc/numpy/reference/distutils_guide.html
```
I'm sure there are a lot of these. I changed some important ones.

mdhaber commented on doc/source/dev/contributor/benchmarking.rst:
```diff
@@ -0,0 +1,210 @@
+submitting a pull request.
+
+To run all benchmarks, navigate to the root SciPy directory at the
+command line and execute
```
Done

mdhaber commented on doc/source/conf.py:
```diff
@@ -104,16 +104,16 @@
-    'dev/releasing.rst',
-    'dev/versioning.rst',
+    # these are all included directly in dev/core-dev/index.rst:
+    'dev/core-dev/decisions.rst',
```
After renaming the files, removing the exclude patterns, modifying the include directives, e.g.
```
.. include:: decisions.rst.inc
```
and waiting for it to rebuild from scratch, I found `/dev/core-dev/index.html` to be empty. Had to revert.

mdhaber replied:
Saw your [PR](https://github.com/mdhaber/scipy/pull/1/files) so I tried again.

The problem was that when I attempted to change the filename extension, macOS  appended `.rst` onto it again and then automatically turned on "Hide filename extension" for that file. So my files were actually `.rst.inc.rst` but it only showed me `.rst.inc`, as though I had successfully changed the extension. 

No `./` because all the files are in the same directory as `index.rst`.

mdhaber commented:
@fletcheaston please review changes and "resolve conversations" as appropriate.
@larsoner Made most changes suggested and deferred two to #10366.

Kai-Striega commented on doc/source/building/faq.rst:
```diff
@@ -23,7 +23,7 @@ command ``source scipy-dev/bin/activate``, and ``deactivate`` to exit from the
 activated, install first Scipy's dependencies::
 
-    $ pip install NumPy pytest Cython
+    $ pip install NumPy pytest Cython pybind11
```
Looks good to me 👍. Just a minor point, maybe write all of them as either lowercase (prefered) or camel case?

Kai-Striega commented:
@mdhaber I'm happy with the addition of pybind11. I'll leave #10355 open for now, feel free to close it once this is merged.

Kai-Striega commented on doc/source/dev/contributor/quickstart_ubuntu.rst:
```diff
@@ -0,0 +1,113 @@
+
+*Consider watching the companion video* `Anaconda SciPy Dev: Part I (macOS)`_ *before starting. Even though the video is intended for macOS users, most of the procedures are the same or very similar.*
+
+#. Download, install, and test the latest release of the `Anaconda Distribution of Python`_. In addition to the latest version of Python 3, the Anaconda distribution includes dozens of the most popular Python packages for scientific computing, the Spyder integrated development environment (IDE), the ``conda`` package manager, and tools for managing virtual environments.
```
Should we be documenting Anaconda by default?

Anaconda includes quite a few packages not used in building SciPy and leads to a reasonable large download, which may be a barrier to users with slow internet connections. See this [NumPy mailing list](http://numpy-discussion.10968.n7.nabble.com/how-much-does-binary-size-matter-td46885.html#a46886) discussion. Note that they are discussing a **16mb** download. The Anaconda installer on its own is ~650mb.

Second, when using the ``apt`` machinery, a user who can't (or doesn't want to) use Anaconda must specify the requires packages individually. In that case having a reference of the required packages will be useful (plus whenever I rebuild SciPy I can simply copy and paste the requirements)

mdhaber replied:
1. This tutorial is for the average beginning developer. Since the smallest hard drive one can easily buy these days is hundreds of GB and sufficiently large USB flash drives are frequently given away for free, I don't expect the [5GB requirement ](https://docs.anaconda.com/anaconda/install/) to be a problem for the average beginning developer. I'm not saying that this is not a problem for anyone on any platform - I've had a hard time staying under a 2GB quota on a remote server I wanted to use for SciPy development - but that Anaconda serves a wide audience. Moreover, in situations where space is at a premium, these instructions are easily adapted for use with [Miniconda](https://docs.conda.io/en/latest/miniconda.html), which comes with `conda` but without all the other packages. I would be happy to review such a PR, or a PR for any other choice of Python distribution, package manager, virtual environment system, and editor. (The choice of all those is easy if you pick Anaconda, BTW.) At some point we can combine all these to replace the [current build documentation ](https://docs.scipy.org/doc/scipy/reference/building/index.html) entirely, which has never worked for me without modification and (hopefully most would agree) not as easy for a first-time developer to follow as the new guides and videos.

2. Like Step 6 of [Installing SciPy](https://github.com/scipy/scipy/blob/d24db98f364c2a0e498c537ca25d59781f2ad6ed/doc/source/dev/contributor/quickstart_ubuntu.rst#installing-scipy) or do you mean something else?

Kai-Striega commented on doc/source/dev/contributor/quickstart_ubuntu.rst:
```diff
@@ -0,0 +1,113 @@
+
+   In Ubuntu, the Anaconda download comes as a ``.sh`` file. `In a terminal, navigate <https://help.ubuntu.com/community/UsingTheTerminal>`_ to the directory in which the ``.sh`` file is located and enter ``./file.sh``, where ``file.sh`` is to be replaced with the full name of the Anaconda download. This starts the installation process; from there, simply follow the prompts.
+
+#. Install git. An easy way to do this is to enter the command ``sudo apt-get install git`` in the terminal and follow the prompts. We'll use this software to download and manage the SciPy source code.
```
``apt`` will be sufficient rather than the whole ``apt-get``. From the apt man page:

> apt provides a high-level commandline interface for the package management system. It is intended as an end user interface and enables some options better suited for interactive usage by default compared to more specialized APT tools like apt-get(8) and apt-cache(8).

Although I admit this is a *very* minor.

mdhaber replied:
Done.

Kai-Striega commented on doc/source/dev/contributor/quickstart_ubuntu.rst:
```diff
@@ -0,0 +1,113 @@
+
+#. In the terminal, navigate into the ``scipy`` root directory (e.g. ``cd scipy``).
+
+#. Install `Homebrew on Linux`_. Enter into the terminal |br| ``sh -c "$(curl -fsSL https://raw.githubusercontent.com/Linuxbrew/install/master/install.sh)"`` |br| or follow the installation instructions listed on the `Homebrew on Linux`_ website. Homebrew requires additional packages to be installed, and lists the required commands in the terminal window. Copy and paste these commands into the te ...
```
Using Homebew on Linux doesn't really make sense to me, is there a reason we require *"Homebew on Linux"* to install gcc?  Were already installing git via apt. Installing gcc with apt is simply: ``sudo apt install gcc``.

mdhaber replied:
@fletcheaston tried that and that version of `gcc` seemed not to work for him. We tried the Homebrew version and it did work.

Kai-Striega replied:
Ok. From memory I needed to install g++ individually.

mdhaber commented:
@kai-striega @larsoner I've responded to comments and cleaned commit history. Can we merge before the sprint at the SciPy conference (July 13/14)?

Kai-Striega commented:
Sounds good to me.

On Mon., 1 Jul. 2019, 08:33 Matt Haberland, <notifications@github.com>
wrote:

> @Kai-Striega <https://github.com/Kai-Striega> @larsoner
> <https://github.com/larsoner> I've responded to comments and cleaned
> commit history. Can we merge before the sprint at the SciPy conference
> (July 13/14)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/10286?email_source=notifications&email_token=AHFOJNIFRBSBIC2NH6UDTBTP5FGEZA5CNFSM4HWPVCI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY4XAMQ#issuecomment-507080754>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHFOJNMF2X43PZM4HSJQFFTP5FGEZANCNFSM4HWPVCIQ>
> .
>

pvanmulbregt commented:
I added gh-10386 with instructions on adding a new `scipy.stats` distribution.  I kept it separate so that it wouldn't get lost in the conversation of this PR.  Similarly the process to add a function to `scipy.special` is sufficiently complicated that it should be written down.  (There are some old minimal docs in `scipy/doc/source/dev/modules.rst` but developers would probably appreciate step-by-step instructions for the ~10 steps needed.)

Does this Overhaul have an obvious place for such instructions?

mdhaber commented:
@pvanmulbregt adding those is a great idea. There is definitely a good place for them, but yeah - this PR is already massive, so let's get this merged and then you can fit these documents into the new table of contents.

@kai-striega @larsoner I tried to clean the commit history after responding to your comments. Look ready to go? If so, would one of you hit the green button?

Kai-Striega commented:
@mdhaber just confirming merge or squash & merge?

mdhaber commented:
@Kai-Striega I've tried to keep the commits logical. If you think they look OK to you, regular merge is good.

WarrenWeckesser commented:
@mdhaber, a belated *thanks* for this!

Earlier in the discussion you asked about the file `HACKING.rst.txt`.  I still have a question about it.  I'd like to see that file changed to `CONTRIBUTING.rst` (and merge anything from the existing `CONTRIBUTING.rst` into that file if it is not already there).  Having both is clearly redundant, and the current `CONTRIBUTING.rst` has just a couple paragraphs.  The first section title in `HACKING.rst.txt` is ***"Ways to contribute"***, so the name `CONTRIBUTING.rst` seems the most appropriate for the merged file, and these days it just seems like the right name to use.

Is there any reason to *not* merge these?  I can do it if there is no reason not to.

mdhaber commented:
@WarrenWeckesser note that @rgommers wrote on June 10:
>> Why does HACKING.rst.tst exist separately in the scipy root directory? Why not just put the contents in \doc\source\hacking.rst? Just curious.

>It predates more extensive html docs, and that file name is (used to be) commonly used by projects for developer docs.

>> link to CONTRIBUTING.rst somewhere. (I don't know that I've run across this before, so if it had a link before perhaps it was buried?)

>I don't think you should link back to that. It's a very concise intro, and is a file that GitHub recognizes so it needs to be there. It's an intro point to link to the real docs, no need to reverse that flow of information/attention.

If it's OK to break from history and not have a `HACKING.rst.txt`, it sounds like putting that file's contents into `CONTRIBUTING.rst` would satisfy GitHub, and it makes more sense to me, too.
